Add Matter lock event changed_by

jvmahon opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR adds support for the lock "changed_by" attribute.

Adding this attribute support allows the triggering of automation based on how it was operated (Changed by). For example, an automation can now turn on lights when a door is unlocked, but only when it is manually unlocked or by keypad PIN, but not remote.

Note that the addition of lock event handling is the foundation for adding "jammed"lock detection - that will be done in a separate PR.

<img width="602" height="683" alt="image" src="https://github.com/user-attachments/assets/d6ec1f75-391e-4c3e-8c98-57d0fe7656be" />


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jvmahon committed:
Add event support and related attributes

jvmahon committed:
Add lock changed_by and user index attributes

Apollon77 commented:
Matter wise the PR makes sense, I can not judge the Python side of things tbh

MartinHjelmare commented on homeassistant/components/matter/lock.py:
```diff
@@ -41,6 +61,72 @@ class MatterLock(MatterEntity, LockEntity):
+        return self._attr_user_index
+
+    @property
+    def extra_state_attributes(self) -> Mapping[str, Any] | None:
```
We shouldn't add extra state attributes in the lock entity. Add sensor entities instead for these features. One sensor for the user index and one for the changed by.

jvmahon replied:
> We shouldn't add extra state attributes in the lock entity. Add sensor entities instead for these features. One sensor for the user index and one for the changed by.

I would like to see you make an exception here and not move this to an additional sensor. I do understand the issue that extra state attributes can increase database size, but here there are good reasons to add it here (and since locks are not often operated, there should be little impact on the database size). Here are the reasons:

For the "changed_by", this is not being added as by the extra_state_attributes, but is already a "standard" attribute of the lock entity. See the standard "changed_by" property here: https://developers.home-assistant.io/docs/core/entity/lock, so I believe I am using and setting this in the intended manner.

For the "user_index":
 - I would like this data to be available in the trigger.to_state state data passed to automations when a lock is operated so it can be read in automation without having to reference an additional sensor. This makes it significantly easier to write a automations where one of multiple locks can trigger to start the automation, where you want to then look at the changed_by and user_index to decide other operations when the lock triggers.
 - Given the low number of expected lock actuations in a day, the addition to the state data should be quite low. As a rough guess, I'd say a dozen lock operations a day might be a reasonable number - not a lot of extra data.
 - Note that the current event.py file does not support lock events, so there is also an issue there.
 - the lock UI already shows "changed_by" in the lock dialog window. I think its more consistent from a UI view to have the user_index information displayed in the same interface as changed_by so you can view it consistently. As shown in the image from my first post, it seems a better UI to be able to see changed by: keypad, user_index:5. 
 - in the future, I plan to expand this to display the user name which is also data stored with a Matter lock PIN to also allow automation conditions based user name, which I don't think can be done by a sensor, so this is consistent with that future.
 - 
 So, for these reasons, I do think this is a good example of where extra_state_attributes is a good implementation choice.

jvmahon committed:
Merge branch 'home-assistant:dev' into add_lock_event_attributes

jvmahon committed:
Update lock to enable changed_by attribute.

jvmahon committed:
Update lock.py to support changed_by event

MartinHjelmare commented on homeassistant/components/matter/lock.py:
```diff
@@ -41,6 +58,57 @@ class MatterLock(MatterEntity, LockEntity):
+            return
+
+        LOGGER.debug(
+            "Received _on_matter_node_event: event type %s, event id %s for %s with data %s",
```
```suggestion
            "Received node event: event type %s, event id %s for %s with data %s",
```

MartinHjelmare commented on homeassistant/components/matter/lock.py:
```diff
@@ -41,6 +58,57 @@ class MatterLock(MatterEntity, LockEntity):
+        )
+
+        # handle the DoorLock events
+        nodeEventData: dict[str, int] = nodeEvent.data or {}
```
Please use snake_case variable names.

MartinHjelmare commented on homeassistant/components/matter/lock.py:
```diff
@@ -41,6 +58,57 @@ class MatterLock(MatterEntity, LockEntity):
+    def _on_matter_node_event(
+        self,
+        event: EventType,
+        nodeEvent: MatterNodeEvent,
```
```suggestion
        node_event: MatterNodeEvent,
```

MartinHjelmare commented on homeassistant/components/matter/lock.py:
```diff
@@ -41,6 +58,57 @@ class MatterLock(MatterEntity, LockEntity):
+                clusters.DoorLock.Events.LockOperationError.event_id
+            ):  # Lock cluster event 3
+                # if an operation error occurs, clear the optimistic state
+                self._reset_optimistic_state(write_state=True)
```
True is the default.
```suggestion
                self._reset_optimistic_state()
```

MartinHjelmare reviewed (commented):
Please add tests of the new code. Test that the changed by and reset of the optimistic state update works when receiving node events.

jvmahon committed:
Update homeassistant/components/matter/lock.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jvmahon committed:
Update homeassistant/components/matter/lock.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jvmahon committed:
Update homeassistant/components/matter/lock.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jvmahon committed:
Added tests

jvmahon committed:
Add changed_by attribute to Matter lock entity

jvmahon commented:
> Please add tests of the new code. Test that the changed by and reset of the optimistic state update works when receiving node events.

I have added a new test for the "changed_by" attribute. 

I have removed the "case" statement for the reset of the optimistic state, so no test is needed.  Once this PR is accepted, I will be posting another PR (this week) that adds support for the jamming event. In that PR, I will re-add the LockOperationEvent event case and put appropriate test.

jvmahon commented:
> You don't need to know if the lock is locked or unlocked to understand the user index, if the user index represents the last user that changed lock.

I agree that you don't need to know locked or unlocked to understand the user index. But it is also true that the user_index alone does not represent the last user that changed the lock and you can get the same user_index for different users. This is because user_index for a user is not guaranteed to be consistent across the different changed_by values. So, to user user_index you must first test for specific changed_by conditions (if changed_by = keypad the if user_index = XX ...)

In any case, the changed_by alone is still useful, so let's just move ahead. thanks.

MartinHjelmare commented:
Ok. Let's finish this PR with the current approach and try to add user name later, when that is available.

jvmahon commented:
> Ok. Let's finish this PR with the current approach and try to add user name later, when that is available.

Thank you.

MartinHjelmare reviewed (approved):
Thanks!
